lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -287,24 +400,382 @@ withNetworkLayer tr network np project k =
+                        then either (error . show) Just $ fromBlockfrost b
+                        else Nothing
+
+            getNextBlocks :: ChainPoint -> IO (Maybe [Block])
+            getNextBlocks = runBF . \case
+                ChainPointAtGenesis ->
+                    Just . pure <$> fetchGenesisBlock network
+                ChainPoint _slotNo hash ->
+                    -- Only one block is fetched for now, even though the type
+                    -- allows for a list of blocks
+                    Just . pure <$> fetchBlock network hash
+
+            getAddressTxs
+                :: BlockHeader
+                -> BlockHeader
+                -> Either Address RewardAccount
+                -> IO ChainEvents
+            getAddressTxs bhFrom bhTo = runBF . fmap fromBlockEvents . \case
+                Left address -> do
+                    txs <- BF.allPages \paged -> BF.getAddressTransactions'
+                        (BF.Address (encodeAddress @nd address))
+                        paged
+                        BF.Ascending
+                        (Just $ headerToIndex bhFrom)
+                        (Just $ headerToIndex bhTo)
+                    forConcurrently txs \BF.AddressTransaction{..} -> do
+                        (bftx, tx) <-
+                            fetchTransaction network _addressTransactionTxHash
+                        txIndex <-
+                            _addressTransactionTxIndex
+                                <?#> "_addressTransactionTxIndex"
+                        txBlockEvents
+                            bftx
+                            (unsafeMkSublist [(txIndex, tx)])
+                            (wholeList [])
+                Right account -> do
+                    let address = BF.Address $ encodeStakeAddress @nd account
+                    rs <- BF.getAccountRegistrations address <&> fmap \reg ->
+                        ( BF._accountRegistrationTxHash reg
+                        , pure case BF._accountRegistrationAction reg of
+                            BF.Registered -> CertRegisterKey account
+                            BF.Deregistered -> CertDelegateNone account
+                        )
+                    ds <- BF.getAccountDelegations address >>= traverse \d -> do
+                        pool <- fromBlockfrostM $ BF._accountDelegationPoolId d
+                        pure ( BF._accountDelegationTxHash d
+                             , [ CertDelegateFull account pool ]
+                             )
+                    blockEventsRegDeleg <-
+                        forConcurrently (rs <> ds) \(txHash, dcerts) -> do
+                        tx@BF.Transaction{_transactionIndex} <- BF.getTx txHash
+                        txIndex <- _transactionIndex <?#> "_transactionIndex"
+                        txBlockEvents
+                            tx
+                            (wholeList [])
+                            (unsafeMkSublist $ (txIndex,) <$> dcerts)
```

Comment:
Alas, using the same index `txIndex` for multiple items does invalidate the internal invariants of `Sublist`. 😔  This may affect `mergeSublist`, which in turn may affect `mergeSameBlock` and give incorrect results in the event that one transaction features multiple delegation certificates *and* creates a transaction output — one of the certificates will be lost when `mergeSameBlock` is applied much later.

In addition, I think that we don't have a guarantee that `dcerts` is ordered correctly — could you use `getTxDelegations` instead to get the delegation certificates in the right order? (The result of `getTxDelegations` may need to be filtered and sorted on the field `_transactionDelegationCertIndex`.)

Perhaps the simplest solution is to reuse `fetchDelegation` here — it's ok to report more data than we would strictly need.